### ISSUE TITLE
idiomatic twisted.web.template.renderElement usage could have greatly lower costs and resulting response size

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#10348]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#10348|
|Type|enhancement|
|Created|2022-06-03 20:17:21Z|

Consider this behavior of `renderElement`:

```
>>> from twisted.web.template import renderElement, XMLString
>>> class Req:
...   def write(self, bs): chunks.append(bs)
...   def finish(self): pass
... 
>>> chunks = []
>>> renderElement(Req(), XMLString("<table><td><tr>Hi</tr></td></table>").load())
1
>>> chunks
[b'<!DOCTYPE html>', b'\n', b'<', b'table', b'>', b'<', b'td', b'>', b'<', b'tr', b'>', b'Hi', b'</tr>', b'</td>', b'</table>']
```

If the response is using chunked transfer encoding then each of these writes incurs 4 bytes of overhead (\r\n\r\n) plus 1 (or sometimes possibly 2 or 3) more for the chunk length.  Since these strings are quite short (many are only 1 byte themselves) this ends up being quite a lot of overhead - as much as 5x for the shortest strings.  This is overhead on the size of the response - this data has to be copied around in memory, sent over the network, and interpreted by the client.

In the case of this simple example, the whole response is known synchronously (ie, before `renderElement` even returns) and there's no reason it couldn't all end up in a single chunk (ie, a single call to Request.write).

For elements containing a Deferred, at least the strings between Deferreds could be combined which still greatly reduces the number of small chunks written (even with every single string coming from a Deferred, it's no _worse_ than the current behavior).

<details><summary>Searchable metadata</summary>

```
trac-id__10348 10348
type__enhancement enhancement
reporter__exarkun exarkun
priority__normal normal
milestone__None None
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__web web
keywords__review review
time__1654287441819996 1654287441819996
changetime__1655256998115097 1655256998115097
version__None None
owner__exarkun exarkun

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

https://github.com/twisted/twisted/pull/1736

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun set status to `closed`|
|-|-|

Fixed by 6f8774cf7a2536a8551635e01118c1eae337d9d6
